File: backend/app/providers/india_mf.py
"""Indian mutual funds via mfapi.in (free, keyless, no rate limit).

Docs: https://www.mfapi.in/
- Latest NAV:  GET /mf/{schemeCode}/latest
- Full NAV:    GET /mf/{schemeCode}          (newest-first history)
- Search:      GET /mf/search?q=

NAVs are published once per business day (see scheduler for cadence rationale).
All values are in INR. Scheme codes are the AMFI codes used as the holding ``key``.
"""
from __future__ import annotations

import asyncio
from datetime import datetime, timezone

import httpx

from ..models import IN_MF
from .base import CandlePoint, Quote, SearchResult

BASE_URL = "https://api.mfapi.in"
CURRENCY = "INR"

# mfapi.in returns dates as dd-mm-yyyy
_DATE_FMT = "%d-%m-%Y"

# Approximate point cap per range (history is daily).
_RANGE_POINTS = {
    "1D": 2,
    "1W": 6,
    "1M": 23,
    "3M": 66,
    "1Y": 260,
    "ALL": 100000,
}


def _parse_date(raw: str) -> datetime:
    return datetime.strptime(raw, _DATE_FMT).replace(tzinfo=timezone.utc)


class IndiaMutualFundProvider:
    async def get_quotes(self, keys: list[str]) -> list[Quote]:
        if not keys:
            return []
        async with httpx.AsyncClient(base_url=BASE_URL, timeout=20) as client:
            results = await asyncio.gather(
                *(self._quote_one(client, k) for k in keys),
                return_exceptions=True,
            )
        return [q for q in results if isinstance(q, Quote)]

    async def _quote_one(self, client: httpx.AsyncClient, key: str) -> Quote:
        resp = await client.get(f"/mf/{key}/latest")
        resp.raise_for_status()
        data = resp.json()
        points = data.get("data") or []
        if not points:
            raise ValueError(f"no NAV for scheme {key}")
        latest = points[0]
        nav = float(latest["nav"])
        prev = float(points[1]["nav"]) if len(points) > 1 else None
        return Quote(
            key=key,
            price=nav,
            currency=CURRENCY,
            prev_close=prev,
            as_of=_parse_date(latest["date"]),
        )

    async def get_candles(self, key: str, range_key: str) -> list[CandlePoint]:
        limit = _RANGE_POINTS.get(range_key, 260)
        async with httpx.AsyncClient(base_url=BASE_URL, timeout=30) as client:
            resp = await client.get(f"/mf/{key}")
            resp.raise_for_status()
            data = resp.json()
        points = data.get("data") or []
        # history is newest-first; take the most recent `limit`, return oldest-first
        recent = points[:limit]
        out = [
            CandlePoint(ts=_parse_date(p["date"]), close=float(p["nav"]))
            for p in recent
        ]
        out.reverse()
        return out

    async def search(self, query: str) -> list[SearchResult]:
        async with httpx.AsyncClient(base_url=BASE_URL, timeout=20) as client:
            resp = await client.get("/mf/search", params={"q": query})
            resp.raise_for_status()
            data = resp.json()
        return [
            SearchResult(
                key=str(item["schemeCode"]),
                display_name=item.get("schemeName", str(item["schemeCode"])),
                asset_type=IN_MF,
                currency=CURRENCY,
            )
            for item in data[:20]
        ]

File: mobile/services/providers/indiaMf.ts
// Indian mutual funds via mfapi.in (free, keyless). Ported from india_mf.py.
import type { RangeKey } from "@/db/types";

import type { CandlePoint, Quote, SearchResult } from "./types";

const BASE = "https://api.mfapi.in";
const CURRENCY = "INR";

const RANGE_POINTS: Record<RangeKey, number> = {
  "1D": 2,
  "1W": 6,
  "1M": 23,
  "3M": 66,
  "1Y": 260,
  ALL: 100000,
};

// mfapi.in dates are dd-mm-yyyy
function parseDate(raw: string): string {
  const [d, m, y] = raw.split("-");
  return new Date(Date.UTC(Number(y), Number(m) - 1, Number(d))).toISOString();
}

export async function getQuote(key: string): Promise<Quote | null> {
  const res = await fetch(`${BASE}/mf/${encodeURIComponent(key)}/latest`);
  if (!res.ok) throw new Error(`mfapi latest ${res.status}`);
  const data = await res.json();
  const points: { date: string; nav: string }[] = data.data ?? [];
  if (points.length === 0) return null;
  const latest = points[0];
  return {
    key,
    price: Number(latest.nav),
    currency: CURRENCY,
    prev_close: points.length > 1 ? Number(points[1].nav) : null,
    as_of: parseDate(latest.date),
  };
}

export async function getQuotes(keys: string[]): Promise<Quote[]> {
  const results = await Promise.allSettled(keys.map((k) => getQuote(k)));
  const out: Quote[] = [];
  for (const r of results) {
    if (r.status === "fulfilled" && r.value) out.push(r.value);
  }
  return out;
}

export async function getCandles(key: string, range: RangeKey): Promise<CandlePoint[]> {
  const limit = RANGE_POINTS[range] ?? 260;
  const res = await fetch(`${BASE}/mf/${encodeURIComponent(key)}`);
  if (!res.ok) throw new Error(`mfapi history ${res.status}`);
  const data = await res.json();
  const points: { date: string; nav: string }[] = data.data ?? [];
  const recent = points.slice(0, limit).map((p) => ({
    ts: parseDate(p.date),
    close: Number(p.nav),
  }));
  recent.reverse(); // oldest-first
  return recent;
}

export async function search(query: string): Promise<SearchResult[]> {
  const res = await fetch(`${BASE}/mf/search?q=${encodeURIComponent(query)}`);
  if (!res.ok) throw new Error(`mfapi search ${res.status}`);
  const data = await res.json();
  return (data as { schemeCode: number; schemeName: string }[])
    .slice(0, 25)
    .map((item) => ({
      key: String(item.schemeCode),
      name: item.schemeName ?? String(item.schemeCode),
      asset_type: "in_mf" as const,
      currency: CURRENCY,
    }));
}
